File: ios/GroupOps.swift
import Foundation
import ExpoModulesCore
import LibSignalClient

// MARK: - Records

struct SenderKeyOpConfig: Record {
  @Field var senderName: String = ""
  @Field var senderDeviceId: UInt32 = 0
  @Field var nowMs: Double = 0
}

struct CreateSenderKeyDistributionResult: Record {
  @Field var message: Data = Data()
  @Field var newRecord: Data = Data()
}

struct ProcessSenderKeyDistributionResult: Record {
  @Field var newRecord: Data = Data()
}

struct GroupEncryptResult: Record {
  @Field var ciphertext: Data = Data()
  @Field var newRecord: Data = Data()
}

struct GroupDecryptResult: Record {
  @Field var plaintext: Data = Data()
  @Field var newRecord: Data = Data()
}

// MARK: - Capturing store

// Captures the post-op SenderKeyRecord for (sender, distributionId). libsignal
// stores after createSKDM / processSKDM / groupEncrypt / groupDecrypt; this
// wrapper reads it back so the JS layer can persist it to the real store.
private final class CapturingSenderKeyStore: SenderKeyStore {
  private var records: [String: SenderKeyRecord] = [:]

  private static func key(_ sender: ProtocolAddress, _ id: UUID) -> String {
    return "\(sender.name).\(sender.deviceId).\(id.uuidString)"
  }

  func storeSenderKey(from sender: ProtocolAddress, distributionId: UUID, record: SenderKeyRecord, context: StoreContext) throws {
    records[Self.key(sender, distributionId)] = record
  }

  func loadSenderKey(from sender: ProtocolAddress, distributionId: UUID, context: StoreContext) throws -> SenderKeyRecord? {
    return records[Self.key(sender, distributionId)]
  }
}

// MARK: - Helpers

private func makeStore(senderAddress: ProtocolAddress, distributionId: UUID, existingRecord: Data?) throws -> CapturingSenderKeyStore {
  let store = CapturingSenderKeyStore()
  if let bytes = existingRecord {
    let record = try SenderKeyRecord(bytes: bytes)
    try store.storeSenderKey(from: senderAddress, distributionId: distributionId, record: record, context: NullContext())
  }
  return store
}

private func uuidOrThrow(_ s: String) throws -> UUID {
  guard let id = UUID(uuidString: s) else {
    throw Exception(name: "LibsignalError", description: "invalid distributionId \(s)")
  }
  return id
}

private func senderAddress(_ config: SenderKeyOpConfig) throws -> ProtocolAddress {
  return try ProtocolAddress(name: config.senderName, deviceId: config.senderDeviceId)
}

private func loadNewRecord(_ store: CapturingSenderKeyStore, _ sender: ProtocolAddress, _ id: UUID) throws -> Data {
  guard let rec = try store.loadSenderKey(from: sender, distributionId: id, context: NullContext()) else {
    throw Exception(name: "LibsignalError", description: "expected store to contain a new SenderKeyRecord after op")
  }
  return rec.serialize()
}

// MARK: - createSenderKeyDistributionOp

func runCreateSenderKeyDistributionOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  existingRecord: Data?
) throws -> CreateSenderKeyDistributionResult {
  let id = try uuidOrThrow(distributionId)
  let sender = try senderAddress(config)
  let store = try makeStore(senderAddress: sender, distributionId: id, existingRecord: existingRecord)
  let skdm = try SenderKeyDistributionMessage(from: sender, distributionId: id, store: store, context: NullContext())
  var result = CreateSenderKeyDistributionResult()
  result.message = skdm.serialize()
  result.newRecord = try loadNewRecord(store, sender, id)
  return result
}

// MARK: - processSenderKeyDistributionOp

func runProcessSenderKeyDistributionOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  message: Data,
  existingRecord: Data?
) throws -> ProcessSenderKeyDistributionResult {
  let id = try uuidOrThrow(distributionId)
  let sender = try senderAddress(config)
  let store = try makeStore(senderAddress: sender, distributionId: id, existingRecord: existingRecord)
  let skdm = try SenderKeyDistributionMessage(bytes: message)
  try processSenderKeyDistributionMessage(skdm, from: sender, store: store, context: NullContext())
  var result = ProcessSenderKeyDistributionResult()
  result.newRecord = try loadNewRecord(store, sender, id)
  return result
}

// MARK: - groupEncryptOp

func runGroupEncryptOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  plaintext: Data,
  existingRecord: Data
) throws -> GroupEncryptResult {
  let id = try uuidOrThrow(distributionId)
  let sender = try senderAddress(config)
  let store = try makeStore(senderAddress: sender, distributionId: id, existingRecord: existingRecord)
  let ciphertext = try groupEncrypt(plaintext, from: sender, distributionId: id, store: store, context: NullContext())
  var result = GroupEncryptResult()
  result.ciphertext = Data(ciphertext.serialize())
  result.newRecord = try loadNewRecord(store, sender, id)
  return result
}

// MARK: - groupDecryptOp

func runGroupDecryptOp(
  config: SenderKeyOpConfig,
  ciphertext: Data,
  existingRecord: Data
) throws -> GroupDecryptResult {
  // distributionId is on the ciphertext, not the config — the in-memory store
  // is keyed by (sender, distributionId) and the existing record was looked up
  // by the JS layer from the same SenderKeyMessage, so we re-derive the id.
  let parsed = try SenderKeyMessage(bytes: ciphertext)
  let sender = try senderAddress(config)
  let id = parsed.distributionId
  let store = try makeStore(senderAddress: sender, distributionId: id, existingRecord: existingRecord)
  let plaintext = try groupDecrypt(ciphertext, from: sender, store: store, context: NullContext())
  var result = GroupDecryptResult()
  result.plaintext = Data(plaintext)
  result.newRecord = try loadNewRecord(store, sender, id)
  return result
}

File: android/src/main/java/expo/modules/libsignal/GroupOps.kt
package expo.modules.libsignal

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.util.UUID
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.groups.GroupCipher
import org.signal.libsignal.protocol.groups.GroupSessionBuilder
import org.signal.libsignal.protocol.groups.state.SenderKeyRecord
import org.signal.libsignal.protocol.groups.state.SenderKeyStore
import org.signal.libsignal.protocol.message.SenderKeyDistributionMessage
import org.signal.libsignal.protocol.message.SenderKeyMessage

class SenderKeyOpConfig : Record {
  @Field var senderName: String = ""
  @Field var senderDeviceId: Int = 0
  @Field var nowMs: Double = 0.0
}

class CreateSenderKeyDistributionResult : Record {
  @Field var message: ByteArray = ByteArray(0)
  @Field var newRecord: ByteArray = ByteArray(0)
}

class ProcessSenderKeyDistributionResult : Record {
  @Field var newRecord: ByteArray = ByteArray(0)
}

class GroupEncryptResult : Record {
  @Field var ciphertext: ByteArray = ByteArray(0)
  @Field var newRecord: ByteArray = ByteArray(0)
}

class GroupDecryptResult : Record {
  @Field var plaintext: ByteArray = ByteArray(0)
  @Field var newRecord: ByteArray = ByteArray(0)
}

// Captures the post-op SenderKeyRecord for (sender, distributionId). libsignal
// stores after createSKDM / processSKDM / groupEncrypt / groupDecrypt; this
// wrapper reads it back so the JS layer can persist it to the real store.
private class CapturingSenderKeyStore : SenderKeyStore {
  private val records = mutableMapOf<String, SenderKeyRecord>()
  private fun key(sender: SignalProtocolAddress, id: UUID) = "${sender.name}.${sender.deviceId}.$id"
  override fun storeSenderKey(sender: SignalProtocolAddress, distributionId: UUID, record: SenderKeyRecord) {
    records[key(sender, distributionId)] = record
  }
  override fun loadSenderKey(sender: SignalProtocolAddress, distributionId: UUID): SenderKeyRecord? {
    return records[key(sender, distributionId)]
  }
}

private fun makeStore(sender: SignalProtocolAddress, id: UUID, existing: ByteArray?): CapturingSenderKeyStore {
  val store = CapturingSenderKeyStore()
  if (existing != null) {
    store.storeSenderKey(sender, id, SenderKeyRecord(existing))
  }
  return store
}

private fun senderAddress(config: SenderKeyOpConfig) =
  SignalProtocolAddress(config.senderName, config.senderDeviceId)

private fun loadNewRecord(store: CapturingSenderKeyStore, sender: SignalProtocolAddress, id: UUID): ByteArray {
  val rec = store.loadSenderKey(sender, id)
    ?: throw IllegalStateException("expected store to contain a new SenderKeyRecord after op")
  return rec.serialize()
}

internal fun runCreateSenderKeyDistributionOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  existingRecord: ByteArray?,
): CreateSenderKeyDistributionResult {
  val id = UUID.fromString(distributionId)
  val sender = senderAddress(config)
  val store = makeStore(sender, id, existingRecord)
  val builder = GroupSessionBuilder(store)
  val skdm = builder.create(sender, id)
  val result = CreateSenderKeyDistributionResult()
  result.message = skdm.serialize()
  result.newRecord = loadNewRecord(store, sender, id)
  return result
}

internal fun runProcessSenderKeyDistributionOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  message: ByteArray,
  existingRecord: ByteArray?,
): ProcessSenderKeyDistributionResult {
  val id = UUID.fromString(distributionId)
  val sender = senderAddress(config)
  val store = makeStore(sender, id, existingRecord)
  val builder = GroupSessionBuilder(store)
  val skdm = SenderKeyDistributionMessage(message)
  builder.process(sender, skdm)
  val result = ProcessSenderKeyDistributionResult()
  result.newRecord = loadNewRecord(store, sender, id)
  return result
}

internal fun runGroupEncryptOp(
  config: SenderKeyOpConfig,
  distributionId: String,
  plaintext: ByteArray,
  existingRecord: ByteArray,
): GroupEncryptResult {
  val id = UUID.fromString(distributionId)
  val sender = senderAddress(config)
  val store = makeStore(sender, id, existingRecord)
  val cipher = GroupCipher(store, sender)
  val ciphertext = cipher.encrypt(id, plaintext)
  val result = GroupEncryptResult()
  result.ciphertext = ciphertext.serialize()
  result.newRecord = loadNewRecord(store, sender, id)
  return result
}

internal fun runGroupDecryptOp(
  config: SenderKeyOpConfig,
  ciphertext: ByteArray,
  existingRecord: ByteArray,
): GroupDecryptResult {
  // distributionId is on the ciphertext, not the config — the in-memory store
  // is keyed by (sender, distributionId) and the existing record was looked up
  // by the JS layer from the same SenderKeyMessage, so we re-derive the id.
  val parsed = SenderKeyMessage(ciphertext)
  val sender = senderAddress(config)
  val id = parsed.distributionId
  val store = makeStore(sender, id, existingRecord)
  val cipher = GroupCipher(store, sender)
  // GroupCipher.decrypt on Android takes the raw bytes (not the parsed message).
  val plaintext = cipher.decrypt(ciphertext)
  val result = GroupDecryptResult()
  result.plaintext = plaintext
  result.newRecord = loadNewRecord(store, sender, id)
  return result
}
